zcm/julia/ZCM.jl

```diff
@@ -15,320 +35,253 @@ end
+                       Ptr{Native.UvSub}, 
+                       (Ptr{Void}, Ptr{Void}),
+                       c_handler, Ref(options))
+    uv_handler = cglobal(("uv_zcm_msg_handler_trigger", "libzcmjulia"))
+    csub = ccall(("zcm_subscribe", "libzcm"), Ptr{Native.Sub},
+                (Ptr{Native.Zcm}, Cstring, Ptr{Void}, Ptr{Native.UvSub}),
+                zcm, channel, uv_handler, uv_wrapper)
+    sub = Subscription(options, c_handler, uv_wrapper, uv_handler, csub)
+    push!(zcm.subscriptions, sub)
+    return sub
 end
-export subscribe;
 
-# RRR: might be a way to make this template a bit prettier
 # TODO: get this into docs:
 # Note to self: handler could be either a function or a functor, so long as it has
-#               handler(rbuf::RecvBuf, channel::String, msg::MsgType, usr) defined
-function subscribe(zcm::Zcm, channel::AbstractString, MsgType::DataType, handler, usr)
-    return subscribe(zcm, channel,
-                     function (rbuf::RecvBuf, channel::AbstractString, usr)
-                         msg = decode(MsgType, rbuf.data)
-                         handler(rbuf, channel, msg, usr);
-                     end, usr);
+#               handler(channel::String, msg::msgtype) defined
+function subscribe(zcm::Zcm, channel::AbstractString, msgtype::Type{<:AbstractZCMType}, handler::Function)
+    subscribe(zcm, channel, SubscriptionOptions(msgtype, handler))
 end
-export subscribe;
 
-function unsubscribe(zcm::Zcm, sub::Sub)
+function unsubscribe(zcm::Zcm, sub::Subscription)
     ret = ccall(("zcm_unsubscribe", "libzcm"), Cint,
-                 (Ptr{Native.Zcm}, Ptr{Native.Sub}), zcm.zcm, sub.sub);
+                 (Ptr{Native.Zcm}, Ptr{Native.Sub}), zcm, sub.native_sub)
     ccall(("uv_zcm_msg_handler_destroy", "libzcmjulia"), Void,
-          (Ptr{Native.UvSub},), sub.uv_wrapper);
-    return ret;
+          (Ptr{Native.UvSub},), sub.uv_wrapper)
+    deleteat!(zcm.subscriptions, findin(zcm.subscriptions, [sub]))
+    return ret
 end
-export unsubscribe;
 
-function publish(zcm::Zcm, channel::AbstractString, data::Array{UInt8}, datalen::UInt32)
+function publish(zcm::Zcm, channel::AbstractString, data::Vector{UInt8})
     return ccall(("zcm_publish", "libzcm"), Cint,
                  (Ptr{Native.Zcm}, Cstring, Ptr{Void}, UInt32),
-                 zcm.zcm, channel, data, datalen);
+                 zcm, convert(String, channel), data, length(data))
 end
-export publish;
 
 # TODO: force msg to be derived from our zcm msg basetype
-function publish(zcm::Zcm, channel::AbstractString, msg)
-    data = encode(msg)
-    return publish(zcm, channel, data, UInt32(length(data)));
-end
-export publish;
+publish(zcm::Zcm, channel::AbstractString, msg) = publish(zcm, channel, encode(msg))
 
-function flush(zcm::Zcm)
-    return ccall(("zcm_flush", "libzcm"), Void, (Ptr{Native.Zcm},), zcm.zcm);
-end
-export flush;
+Base.flush(zcm::Zcm) = ccall(("zcm_flush", "libzcm"), Void, (Ptr{Native.Zcm},), zcm)
```

Comment:
Weird that you have to have an intimate knowledge of which functions in `Base` you are overloading. Should we just `importall Base`? Maybe `import Base.flush` and just call this function `flush`

Reply:
Forgetting to import something from Base isn't the end of the world (as long as you're not relying on it being imported), but I agree that it's a little awkward. I would not recommend `importall` (in fact, it's likely to be eliminated before julia v1.0 because it's not generally good practice). But if you don't like `Base.fush(zcm::...` then I'm happy to do all the imports at the beginning.

Reply:
resolved in 445dbe9